### PR TITLE
Candidate service and controller creating

### PR DESCRIPTION
**Pull request by Evgeniy-Sinicin**

### CODE REVIEW

File: InternshipBE/BL/Services/CandidateService.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using DAL.Entities;
+using DAL.Interfaces;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace BL.Services
+{
+    public class CandidateService : ICandidateService
+    {
+        private IUnitOfWork _uow;
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: fullname

File: InternshipBE/BL/Services/CandidateService.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        public CandidateService(IUnitOfWork uow, IMapper mapper)
+        {
+            _uow = uow;
+            _mapper = mapper;
+        }
+
+        public async Task<CandidateDTO> GetCandidateAsync(int id)
+        {
+            var candidateEntity = await _uow.Candidates.GetByIdAsync(id);
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: rename

File: InternshipBE/BL/Services/CandidateService.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+            await _uow.Save();
+
+            return _mapper.Map<CandidateDTO>(candidate);
+        }
+
+        public async Task<CandidateDTO> UpdateCandidateAsync(CandidateDTO updatedCandidate)
+        {
+            var mappedCandidate = _mapper.Map<Candidate>(updatedCandidate);
+            var resultCandidate = await _uow.Candidates.UpdateAsync(mappedCandidate);
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: rename

File: InternshipBE/DAL/Repositories/CandidateRepository.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public class CandidateRepository : GenericRepository<Candidate>, ICandidateRepository
+    {
+        public CandidateRepository(ApplicationDbContext context) : base(context)
+        {
+        }
+
+        public async Task<List<Candidate>> GetAllByInternshipIdAsync(int id, int itemsCount, int pageNumber)
+        {
+            return await Task.Run(() => _context.Internships.FirstOrDefault(x => x.Id == id)?.Candidate
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/CandidateRepository.cs]: ?

Comment (romanstat) [reply InternshipBE/DAL/Repositories/CandidateRepository.cs]: remove Task.Run

Comment (romanstat) [reply InternshipBE/DAL/Repositories/CandidateRepository.cs]: ToListAsync

File: InternshipBE/WebApi/Models/GetCandidatesByInternshipIdRequest.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Shared.Models
+{
+    public class GetCandidatesByInternshipIdRequest
+    {
+        public int InternshipId { get; set; }
```
Comment (romanstat) [inline InternshipBE/WebApi/Models/GetCandidatesByInternshipIdRequest.cs:5]: spaces

File: InternshipBE/WebApi/Controllers/CandidateController.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        [HttpPost("CreateCandidate")]
+        public async Task<IActionResult> CreateCandidate([FromBody] CandidateDTO candidate)
+        {
+            var createdCandidate = await _service.CreateCandidateAsync(candidate);
+
+            return Ok(createdCandidate);
+        }
+
+        [HttpPut("UpdateCandidate")]
+        public async Task<IActionResult> UpdateCandidate([FromBody] CandidateDTO updatedCandidate)
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/CandidateController.cs]: candidate

File: InternshipBE/WebApi/Controllers/CandidateController.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            var candidate = await _service.GetCandidateAsync(candidateId);
+
+            return Ok(candidate);
+        }
+
+        [HttpPost("GetCandidateListByInternshipId")]
+        public async Task<IActionResult> GetCandidateListByInternshipId([FromBody] GetCandidatesByInternshipIdRequest body)
+        {
+            var candidates = await _service.GetAllByInternshipIdAsync(body.InternshipId,
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/CandidateController.cs]: inline

File: InternshipBE/BL/Services/CandidateService.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+        public CandidateService(IUnitOfWork uow, IMapper mapper)
+        {
+            _unitOfWork = uow;
+            _mapper = mapper;
+        }
+
+        public async Task<CandidateDTO> GetCandidateAsync(int id)
+        {
+            var candidateEntity = await _unitOfWork.Candidates.GetByIdAsync(id);
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: rename to candidate

File: InternshipBE/BL/Services/CandidateService.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+            await _unitOfWork.Save();
+
+            return _mapper.Map<CandidateDTO>(candidate);
+        }
+
+        public async Task<CandidateDTO> UpdateCandidateAsync(CandidateDTO updatedCandidate)
+        {
+            var mappedCandidate = _mapper.Map<Candidate>(updatedCandidate);
+            var resultCandidate = await _unitOfWork.Candidates.UpdateAsync(mappedCandidate);
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: candidate

File: InternshipBE/BL/Services/CandidateService.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+using System.Threading.Tasks;
+
+namespace BL.Services
+{
+    public class CandidateService : ICandidateService
+    {
+        private IUnitOfWork _unitOfWork;
+        private IMapper _mapper;
+
+        public CandidateService(IUnitOfWork uow, IMapper mapper)
```
Comment (romanstat) [inline InternshipBE/BL/Services/CandidateService.cs]: unitOfWork

File: InternshipBE/DAL/Interfaces/IUnitOfWork.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace DAL.Interfaces
+{
+    public interface IUnitOfWork : IDisposable
+    {
+        ICandidateRepository Candidates { get; }
+
+        IInternshipRepository Internships { get; }
+
+        Task Save();
```
Comment (romanstat) [inline InternshipBE/DAL/Interfaces/IUnitOfWork.cs]: SaveAsync

**Review by romanstat — APPROVED** [review]

**Review by romanstat — APPROVED** [review]

**Review by Nikityanov — APPROVED** [review]

**Review by vldmrklnk — APPROVED** [review]

**Pull request merged**